### ISSUE TITLE
Check and update dependencies in GitHub Actions action.yml

### ISSUE THREAD

**Issue by slawekjaranowski** [dependabot/dependabot-core#4178]
When using `composite` actions steps are defined in `action.yml` and we can have dependency to another actions.

Currently dependencies aren't checked in `action.yml`

**Comment by georgettica** [comment]
This issue infers there is already partial support for dependency update for github-actions.. can you direct me to docs on that?

**Comment by slawekjaranowski** [comment]
https://github.community/t/using-dependabot-with-new-composite-actions/198740
https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/keeping-your-actions-up-to-date-with-dependabot

**Comment by localheinz** [comment]
Any news on this one?

Would be great if composite actions could be updated automatically!

**Comment by georgettica** [comment]
@localheinz by the commits and closing of the issue, this seems to be done
@jurre LMK if there is something missing on this

**Comment by briantist** [comment]
@georgettica composite actions (all actions I think) do not have to be in the root of the repository, and a repository may contain multiple actions.

An action can be referenced like `uses: org/repo/directory/otherdir/action-name@v1` for example. 

Composite actions in the same repo can also be referenced as local files, `uses: ./my-action/otherdir/action-name`

Does the given PR handle those? It seems to make assumptions about the location of the `actions.yml`

**Comment by jurre** [comment]
> I don't think this is working.

The change has not been released and rolled out yet, the issue was automatically closed because it was referenced by the commit, I'll reopen it for clarity.

**Comment by Pwd9000-ML** [comment]
@jurre Thank you very much for clarification :)

**Comment by JonasAlfredsson** [comment]
Hi, 

I am the one that created the pull request, and I was unaware that you were allowed to put the `action.yml` in a subfolder. 
The examples I looked at always put it in the root, so I thought this was an obligatory place just like `.github/workflows/`.
So yes, in the pull request I made the assumption that is should always be like that.

If actions allow that much more freedom then I suppose the change will become a bit more complex. 
A suggestion that just popped into my head would perhaps be to check every `.yml` file in the path that is given in dependabot.yml. But that would mean breaking current setups since they would have to be changed to this:

```
  - package-ecosystem: "github-actions"
    directory: "-github/workflows"
    schedule:
      interval: "daily"
```

**Comment by briantist** [comment]
thanks for updating @JonasAlfredsson ; what you contributed covers most use cases for _published_ composite actions, which have a single action in its own repo, and typically put the `action.yml` in the root, but yeah unfortunately for dependabot, the actual usage is much more flexible (workflows however, even reusable workflows, are not flexible and must be in `.github/workflows`).

**Comment by JonasAlfredsson** [comment]
@jurre, should the pull request be reverted since this doesn't cover all possible cases?

If yes, then we probably need to discuss the best way to approach this.

**Comment by JonasAlfredsson** [comment]
@airtower-luna, yeah we could probably change the code to work relative to the directory path, but if I understand @briantist right the action files do not necessarily need to be named `action.yml`, so we would actually have to run through all `*.yml` files in the given directory for this to be more complete?

**Comment by briantist** [comment]
> @airtower-luna, yeah we could probably change the code to work relative to the directory path, but if I understand @briantist right the action files do not necessarily need to be named `action.yml`, so we would actually have to run through all `*.yml` files in the given directory for this to be more complete?

Sorry to mislead, the file must be named `action.yml` but it can be located anywhere.

**Comment by airtower-luna** [comment]
@JonasAlfredsson, [the docs say](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#about-yaml-syntax-for-github-actions):

> The metadata filename must be either `action.yml` or `action.yaml`.

The point was just that this file doesn't have to be at the root of the repository. For example GitHub's own https://github.com/github/codeql-action has multiple subdirectories that each contain an Action (IIRC it's discouraged on the marketplace, though). And with local actions it's super common to have them in a subdirectory of the repository.

So searching for `action.yml` or `action.yaml` in the configured directory (plus workflows as before if it's `/`) would actually be enough to cover all cases. :smiley_cat:

**Comment by JonasAlfredsson** [comment]
While it looks like it will work satisfactorily with a custom directory on the current patch it was not known or handled by me in the tests, so I created another patch here: https://github.com/dependabot/dependabot-core/pull/4815
I hope this will be sufficient :)

**Comment by briantist** [comment]
Tack @JonasAlfredsson ! Your work is appreciated :)

**Comment by JonasAlfredsson** [comment]
Ooooh, it appears to be working :) 
https://github.com/JonasAlfredsson/checkout-qemu-buildx/pull/2

**Comment by fmeriaux** [comment]
For subdirectories it doesn't work.

But I found a workaround: 

```yaml
version: 2
updates:
  # Maintain dependencies for GitHub Actions (/.github/workflows)
  - package-ecosystem: "github-actions"
    directory: "/"
    schedule:
      interval: "daily"

  - package-ecosystem: "github-actions"
    directory: "/actions/<my-custom-action-A>"
    schedule:
      interval: "daily"

  - package-ecosystem: "github-actions"
    directory: "/actions/<my-custom-action-B>"
    schedule:
      interval: "daily"
```

**Comment by georgettica** [comment]
@fmeriaux should you open a card specific with you problem and your workaround?

My guess is that by seeing the problem and your solution it can be solved at the source

**Comment by fmeriaux** [comment]
> @fmeriaux should you open a card specific with you problem and your workaround?
> 
> My guess is that by seeing the problem and your solution it can be solved at the source

The current issue proposes to support composite actions.  When the issue was created this was not the case, since the PR of @JonasAlfredsson they are supported if the action is at the root (of directory configuration). 

It is implied that nested composite actions are supported, but this is not the case. So I can open a "bug" in this sense, unless we stay on this issue that is not "solved" entirely.

**Comment by JonasAlfredsson** [comment]
Hi @fmeriaux ,

I think not traversing into subdirectories is the intended usecase. I do not know if relevant documentation has been updated, but specifying each directory (where an action file is located) is necessary as it is written right now.

While this is perhaps a little bit cumbersome if you have a lot of custom actions, I  feel like this is a safer approach than just capturing every `action.ya?ml` file in either one or unknown levels of subdirectories. However, if the traversing functionality is desired then I do not think it would be _too_ difficult to implement it.

**Comment by fmeriaux** [comment]
> Hi @fmeriaux ,
> 
> I think not traversing into subdirectories is the intended usecase. I do not know if relevant documentation has been updated, but specifying each directory (where an action file is located) is necessary as it is written right now.
> 
> While this is perhaps a little bit cumbersome if you have a lot of custom actions, I feel like this is a safer approach than just capturing every `action.ya?ml` file in either one or unknown levels of subdirectories. However, if the traversing functionality is desired then I do not think it would be _too_ difficult to implement it.

Yes, I agree with your reasoning. It may be more secure to have an explicit intention in the configuration.  

However the documentation is far from clear, I had to experiment and search in your contribution to understand and find a solution to my need.

A good compromise could be to support wildcards, something like: 

```yaml
version: 2
updates:
  # Maintain dependencies for GitHub Actions (/.github/workflows)
  - package-ecosystem: "github-actions"
    directory: "/"
    schedule:
      interval: "daily"

  - package-ecosystem: "github-actions"
    directory: "/actions/**"
    schedule:
      interval: "daily"

```

I can open a new issue for wildcard support, if you want.

**Comment by JonasAlfredsson** [comment]
Two things:

1. I am not sure where documentation is located, and how to update it. Is this something @jurre can answer?
2. I do not have time to look at wildcard matching right now, so please open a ticket and see if someone else in the community has time before me :)

**Comment by fmeriaux** [comment]
Done https://github.com/dependabot/dependabot-core/issues/5137

**Comment by jeffwidman** [comment]
Closing as the overall work has been merged / deployed in https://github.com/dependabot/dependabot-core/pull/4755 and https://github.com/dependabot/dependabot-core/pull/4815 and then the followup idea of not having to specify subdirectories is tracked in #5137.

Many thanks to @JonasAlfredsson for all his hard work here.

**Comment by deivid-rodriguez** [comment]
This is so interesting because I just tried your case and it's indeed not working. But it seems that #6660 just merged has actually regressed this, so this should've been working for you until today. Could it by that your actions were up to date, and that's why no PRs were created? Anyways, I'll be fixing the regression now.

**Comment by saisatishkarra** [comment]
@deivid-rodriguez Yes it seemed to work and create a [PR](https://github.com/Scimia/public-shared-actions/pull/2). When I tried to rebase it dependabot closed the PR and is not generate any new ones and was unable to update dependencies in composite actions. The https://github.com/Scimia/public-shared-actions/blob/main/actions/security-actions/scan-docker-image/action.yaml has old versions of `anchore/scan-action` and `anchore/sbom-action` which needs to be bumped to [v3.3.4](https://github.com/anchore/scan-action/releases/tag/v3.3.4) and [v0.13.3](https://github.com/anchore/sbom-action/releases/tag/v0.13.3) respectively. This seems like a  regression. I will wait until #6690 and see if that works. Thanks for the PR!!!

**Comment by deivid-rodriguez** [comment]
@saisatishkarra I just deployed the fix in #6690, can you try whether that works?

**Comment by saisatishkarra** [comment]
@deivid-rodriguez Thanks for the release.
Dependable doesn't seem to be working in 3 use-cases yet:
1. Update dependencies within a single composite action. 
- Ex: https://github.com/Scimia/public-shared-actions/blob/main/actions/security-actions/scan-docker-image/action.yaml#L71-L123 
- Result: The dependencies are supposed to be updated to `v0.13.3 for anchore/sbom-action` and `v3.3.4 for anchore/scan-action`
- Last known to work: https://github.com/Scimia/public-shared-actions/pull/2 was 3 days ago

2. Update the composite action dependency in a different repository workflow. 
-Ex:  The `uses` block in a workflow (repoA) referring to a composite action in a repoB must be updated when referenced using SHA, tag.

**Comment by deivid-rodriguez** [comment]
Regarding the first issue, I think it should be working. Can you try manually checking for updates at the "Insights > Dependency Graph > Dependabot" tab?

Regarding the second problem, can you open a separate issue with steps to reproduce?

Thanks!

**Comment by saisatishkarra** [comment]
Manually triggering the checking for updates from the insights tab seemed to have **worked** and updated the PRs. I guess we can mark it resolved and dependable to automatically pick up any new updates later down the line.

<img width="1672" alt="Screen Shot 2023-02-21 at 9 47 45 AM" src="https://user-images.githubusercontent.com/29549979/220392984-70732bc3-55cb-4e4b-981a-18cfe9368386.png">

I have filed a [bug](https://github.com/dependabot/dependabot-core/issues/6656) previously for the second problem but this was supposed to be addressed from #6690 and #6660. I don't think this is still working.  Appreciate someone taking a look and if I am missing something. 

Thanks again for the prompt response.

**Comment by deivid-rodriguez** [comment]
Oh, sorry I missed that other issue you had opened. I will check that again and let you know exactly what to expect there.
